sound/soc/sof/intel/hda-dai-ops.c

```diff
@@ -145,8 +145,72 @@ static struct hda_dai_widget_dma_ops hda_dma_ops = {
 	.stream_put_dma = hda_stream_put_dma,
 };
 
+static int hda_codec_dai_params(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+				void *dma_data, struct snd_pcm_substream *substream,
+				struct snd_pcm_hw_params *params, unsigned int *format_val)
+{
+	struct hdac_ext_stream *hext_stream = dma_data;
+	struct hdac_ext_link *hlink;
+	struct hdac_bus *bus;
+	unsigned char stream_tag;
+	unsigned int link_bps;
+
+	bus = sof_to_bus(sdev);
+	hlink = snd_hdac_ext_bus_get_hlink_by_name(bus, codec_dai->component->name);
+	if (!hlink)
+		return -EINVAL;
+
+	stream_tag = hdac_stream(hext_stream)->stream_tag;
+	if (hext_stream->hstream.direction == SNDRV_PCM_STREAM_PLAYBACK)
+		snd_hdac_ext_bus_link_set_stream_id(hlink, stream_tag);
+
+	/* set the hdac_stream in the codec dai */
+	snd_soc_dai_set_stream(codec_dai, hdac_stream(hext_stream), substream->stream);
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+		link_bps = codec_dai->driver->playback.sig_bits;
+	else
+		link_bps = codec_dai->driver->capture.sig_bits;
+
+	*format_val = snd_hdac_calc_stream_format(params_rate(params), params_channels(params),
+						  params_format(params), link_bps, 0);
+
+	dev_dbg(bus->dev, "format_val=%d, rate=%d, ch=%d, format=%d\n",
+		*format_val, params_rate(params), params_channels(params), params_format(params));
+
+	return 0;
+}
+
+static int hda_codec_dai_free(struct snd_sof_dev *sdev, struct snd_soc_dai *codec_dai,
+			      void *dma_data, struct snd_pcm_substream *substream)
+{
+	struct hdac_ext_stream *hext_stream = dma_data;
+	struct hdac_ext_link *hlink;
+	struct hdac_bus *bus;
+
+	bus = sof_to_bus(sdev);
+	hlink = snd_hdac_ext_bus_get_hlink_by_name(bus, codec_dai->component->name);
+	if (!hlink)
+		return -EINVAL;
+
+	if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
+		unsigned char stream_tag;
+
+		stream_tag = hdac_stream(hext_stream)->stream_tag;
+		snd_hdac_ext_bus_link_clear_stream_id(hlink, stream_tag);
+	}
+
+	return 0;
```

Comment:
....but here this is completely unrelated to codec dais. The problem is to find the hlink, and we could use the CPU dai name for this. There's no reason why we should use the codec name, this clearly doesn't translate well for other platforms.

Reply:
the comment still stands.